src/app/modules/form/public-form/public-form.controller.ts

```diff
@@ -350,3 +360,91 @@ export const handleGetPublicForm: RequestHandler<
+ * @param isPersistentLogin whether the client wants to have their login information stored
+ * @returns 200 with the redirect url when the user authenticates successfully
+ * @returns 400 when there is an error on the authType of the form
+ * @returns 400 when the eServiceId of the form does not exist
+ * @returns 404 when form with given ID does not exist
+ * @returns 500 when database error occurs
+ * @returns 500 when the redirect url could not be created
+ * @returns 500 when the redirect feature is not enabled
+ */
+export const _handleFormAuthRedirect: RequestHandler<
+  { formId: string },
+  RedirectUrlDto | ErrorDto,
+  unknown,
+  Query & { isPersistentLogin?: boolean }
+> = (req, res) => {
+  const { formId } = req.params
+  const { isPersistentLogin } = req.query
+  const logMeta = {
+    action: 'handleFormAuthRedirect',
+    ...createReqMeta(req),
+    formId,
+  }
+  // NOTE: Using retrieveFullForm instead of retrieveForm to ensure authType always exists
+  return FormService.retrieveFullFormById(formId)
+    .andThen((form) => {
+      switch (form.authType) {
+        case AuthType.MyInfo:
+          return validateMyInfoForm(form).andThen((form) =>
+            MyInfoFactory.createRedirectURL({
+              formEsrvcId: form.esrvcId,
+              formId,
+              requestedAttributes: form.getUniqueMyInfoAttrs(),
+            }),
+          )
+        case AuthType.SP:
+        case AuthType.CP: {
+          // NOTE: Persistent login is only set (and relevant) when the authType is SP.
+          // If authType is not SP, assume that it was set erroneously and default it to false
+          return validateSpcpForm(form).andThen((form) => {
+            const target = getRedirectTarget(
+              formId,
+              form.authType,
+              isPersistentLogin,
+            )
+            return SpcpFactory.createRedirectUrl(
+              form.authType,
+              target,
+              form.esrvcId,
+            )
+          })
+        }
+        // NOTE: Only MyInfo and SPCP should have redirects as the point of a redirect is
+        // to provide auth for users from a third party
+        default:
+          return err<never, AuthTypeMismatchError>(
+            new AuthTypeMismatchError(form.authType),
+          )
+      }
+    })
+    .map((redirectURL) => {
+      return res.status(StatusCodes.OK).json({ redirectURL })
+    })
+    .mapErr((error) => {
+      logger.error({
+        message: 'Error while creating redirect URL',
+        meta: logMeta,
+        error,
+      })
+      const { statusCode, errorMessage } = mapFormAuthRedirectError(error)
+      return res.status(statusCode).json({ message: errorMessage })
+    })
+}
+
+/**
+ * Handler for /forms/:formId/auth/redirect
+ */
+export const handleFormAuthRedirect = [
+  celebrate({
+    [Segments.QUERY]: Joi.object({
+      isPersistentLogin: Joi.boolean().optional(),
+    }),
```

Comment:
should we validate for the form ID in `Segments.PARAMs` as well? I know we already validate using Express route  matching, but it's nice to have all the validation in one place

Reply:
oh sure! just to make sure we're on the same wavelength, the express regex should still remain also right?

Reply:
yep